Add discussions support

3urobeat opened the pull request:
Hey!  
This PR adds support for scraping information and comments from discussions, as well as posting & deleting comments, subscribing & unsubscribing to/from discussions and setting the amount of comments shown per page.  

I have added a new class which represents an existing discussion supplied via URL and scrapes various information from the DOM:  
- `id` (string) - ID of the discussion, is used in the URL
- `type` (EDiscussionType) - Where the discussion is associated to (app, group or the Steam forum)
- `appID` (string) - The appID this discussion is associated to, if of type EDiscussionType.App
- `forumID` (string) - ID of the forum the discussion is in, is used in the URL
- `gidforum` (string) - ID used in post requests when interacting with discussions, not sure why
- `topicOwner` (string) - ID used in post requests when interacting with discussions, not sure why
- `author` (SteamID) - SteamID object of the author of this discussion
- `postedDate` (Date) - When this discussion was created
- `title` (string) - Title of the discussion
- `content` (string) - Content of the discussion
- `commentsAmount` (number) - Amount of comments present in this discussion
- `answerCommentIndex` (number) - If this discussion has an answer, this is the index of the comment

The functions already mentioned above are implemented in `components/discussions.js`.  
Beside the usual functions for commenting and subscribing, it includes another scraper for retrieving a range of comments. Each comment includes the following information:
- `index` (number) - Index of this comment in the comment chain of this discussion
- `commentId` (number) - ID of this comment, issued by Steam
- `commentLink` (string) - Link to this specific comment, will highlight it
- `authorLink` (string) - Profile link of the author of this comment. I did not convert this one to a SteamID to reduce the large amount of extra requests needed to resolve each author link when requesting a large-ish range
- `postedDate` (Date) - When this comment was posted
- `content` (string) - Content of all comments quoted in this comment and the comment itself

I have not added support for posting discussions as a class only represents one discussion, not a forum.

Test code to get 3 comments from the provided discussion and see the nested quotes support:
```js
community.getSteamDiscussion("https://steamcommunity.com/app/739630/discussions/0/1750150652078713439/", (err, res) => {
    res.getComments(35, 37, (err, comments) => {
        console.log(comments);
    });
});
```

Note: I have encountered an error with the scraper yesterday at discussion.js line 39 getting children of undefined, which I was sadly unable to reproduce again. I just wanna note that the scraper might not be completely failsafe, this is however the same case with the other scrapers as well. If I find any bugs I'll ofc look into fixing them and open another PR, just like previously.

Thanks for reading and I hope you find this feature useful.  
Feel free to use the text from this PR in any documentation page if you wish.  
Have a nice day :)

3urobeat committed:
Add discussion class with data scraper

3urobeat committed:
Scrape comment id marked as answer

3urobeat committed:
Add discussion comment scraping support

3urobeat committed:
Add discussion subscribing support

3urobeat committed:
Add discussion commenting support

3urobeat committed:
Scrape gidforum & topicOwner

3urobeat committed:
Add discussion object methods

3urobeat committed:
Get id more reliably and misc

3urobeat committed:
Add missing getComments() object method and fix callback jsdoc

3urobeat committed:
Fix first comment getting blockquote from another comment

3urobeat committed:
Add type detection and support forum & group discussions

3urobeat committed:
Add setDiscussionCommentsPerPage()

3urobeat committed:
Add support for displaying multiple nested quotes

3urobeat commented:
Oh and while we are at it, I think I mixed [both of these a bit](https://reddit.com/r/ProgrammerHumor/s/hSQCSmpxZX).  
Not terrible, however a uniform usage of either of those would probably be better

Revadike commented:
You prob want to merge it into https://github.com/DoctorMcKay/node-steamcommunity/pull/230

3urobeat commented:
Yeah I thought about that but ultimately decided against it as DoctorMcKay merged my previous PRs into v4 afterwards anyway and the v4 branch seems a bit stale.  
I can open another PR/rebase this one though if you wish

3urobeat committed:
Fix checking for a rejected request

3urobeat committed:
Use eresultError() helper

DoctorMcKay commented:
I'm cool with scraping discussions to retrieve data, but posting threads and comments isn't something I'm comfortable with adding due to the potential for abuse/spam. I'm aware that it's not really difficult for spammers to post spam on their own without the help of the lib, but still. I like to try to maintain at least some plausible deniability for enabling bad people to do bad things.

Revadike commented:
> I'm cool with scraping discussions to retrieve data, but posting threads and comments isn't something I'm comfortable with adding due to the potential for abuse/spam. I'm aware that it's not really difficult for spammers to post spam on their own without the help of the lib, but still. I like to try to maintain at least some plausible deniability for enabling bad people to do bad things.

Valve wouldn't know it came from your library, though. Still, you can combat it somewhat by implementing some guards that align with the [steam discussion rules](https://help.steampowered.com/en/faqs/view/6862-8119-C23E-EA7B), like a blocklist and URL detection. Let's be real, anything can be abused, I hate to see a feature gone just because of theoretical bad actors.

3urobeat commented:
I can understand leaving out voting in the sharedfile PR from a while ago but when the lib already supports posting comments to profiles, groups and sharedfiles then I don't really see where posting comments to discussions is much different.  
In the end it is the user's account that is displayed as the comment author and is responsible for its content. If one wants to post malicious content they could do the same manually - we are not circumventing anything.

> Also, if you're going to target v4 (which seems appropriate), please make sure you're addressing the new lint rules

Yeah, I saw the check yesterday and wanted to address those today

3urobeat committed:
Buff eslint's happiness by 80%

3urobeat committed:
Return 'Discussion Not Found' error when breadcrumbs are missing to avoid unhandled exceptions

3urobeat commented:
I have fixed all eslint errors/warnings. The check is still failing because of other files not relevant to this PR.  
I'm not yet sure how the new `httpRequest()` usage should look like, so I might have to revisit that some time later.

3urobeat committed:
Rename discussion postComment() -> comment()

3urobeat committed:
Update httpRequest helper usage to v4

DoctorMcKay commented:
> I have fixed all eslint errors/warnings. The check is still failing because of other files not relevant to this PR.

Sounds good, I should probably try to fix all those errors in the v4 branch shortly.

> Edit: I have updated the httpRequest() usage - I hope it is correct like this. I briefly tested the methods and they seem to work as expected, I however don't know if I need to do more checking if the request failed (previously the `httpRequestPost` helper returned an `err` param besides the body itself) as I'm currently blindly using the resulting body. Or maybe I'm overthinking this and the `httpRequest` helper already takes care of it?

It's not final yet, but I think the helper is going to take care of all this like it did previously, and reject in cases where it would return an error previously.

Is this ready for review?

3urobeat commented:
Yeah, I think it is ready for review.  
If I notice any oversights or issues I'll of course look into them and open another PR.  
As of now all functions seem work as expected

3urobeat committed:
feat(Discussions): Add eventcomments support

3urobeat committed:
feat(Discussions): Add accountCanComment prop

3urobeat commented:
I've added support for eventcomments too, this is another type of discussion, usually used to post patch notes or make event announcements.  
I have also added a `accountCanComment` prop, because devs can choose if limited accounts are allowed to comment on these posts (for example CS2 & TF2 allow limited accounts, PUBG doesn't - all three are free games).  

I have also noticed that testing the changes made in this PR just by checking out this branch does not work because it was rebased, and therefore `httpRequestGet` still has the old implementation. Just for your information when reviewing.

3urobeat committed:
fix(Discussions): Fix discussion authorLink failing to resolve due to page layout update
